### ISSUE TITLE
[RFC]: Automatic test target determination for CI

### ISSUE THREAD

**Issue by khluu** [vllm-project/vllm#39884]
### Motivation.

Our current conditional testing strategy in CI, which relies on `source_file_dependencies` being manually updated, gets old and inaccurate very quickly. We need to start finding an approach to automatically determine which test to run based on code diff that can adjust and scale with changes, at the same time reduce CI cost by not launching more tests unnecessarily.

### Proposed Change.

I propose that we start with a simple LLM-based approach: an agent instructions markdown file that is integrated with CI pipeline generator

If this approach doesn't work, we can explore more advanced technique: static analysis on dependency graph, train ML model, code coverage report mapping of tests and source files, etc...)

### CC List.

@dougbtv @avinashsingh77 @AndreasKaratzas 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by khluu** [comment]
For feedbacks, we can set up a bot to comment in the PRs with the list of tests that it thinks should be launched. We can collect feedbacks from contributors this way.

**Comment by avinashsingh77** [comment]
I have put a proposal using a LLM based approach where LLM will be fed with the following:
1. PR file diff
2. Mapping of source directories to test directories(derived from AST-based import analysis from [build_test_mapping.py](https://github.com/avinashsingh77/vllm/blob/auto-test-target-determination/.buildkite/scripts/build_test_mapping.py) script). Will look like [test_mapping_output.md](https://github.com/avinashsingh77/vllm/blob/auto-test-target-determination/.buildkite/test_mapping_output.md)
3. An instructions file(includes static rules for cases the mapping can't cover and other instructions)

The LLM will reply with `<test_path> | <reason>`.
To begin with, output will only be posted as a comment in the same PR for the author to provide feedback on tests selection.
Once we gain confidence and iteratively update instructions as per feedback, we will use the LLM output to trigger tests in CI.

Design Document: https://docs.google.com/document/d/1QwVPunKXj8CcuyjS91Rc1R0NYyUosKYz5vXvKFfvpHc 
Please review and comment in the doc with your suggestions.

**Comment by AndreasKaratzas** [comment]
My main concern with the pure LLM approach is that it's absolutely beyond any observability. It'll most likely work, but we'll never know when or how it failed. One way to mitigate that is to go with the agentic setup you proposed and have it really detail the set of tests it considered at the very beginning and how that list evolved through its reasoning. The other thing is to refresh its memory weekly or even more often, so we pseudo-build the dependency graph you described over time.

The other concern I had (and this isn't just an LLM problem, it applies to any heuristic) is how do we account for new changes in the source code? Are we going to have the LLM run across the entire repo after huge PRs? How is its memory going to stay up to date?

Btw, I'm also trying to prototype a similar idea. I was thinking of an LLM, but only for generating embedding vectors for each function/module in vLLM. Then we can train a scoring mechanism on top. For example, suppose the diff in a PR is:

```
file_1, file_2 -> ⟨emb_vec_1, emb_vec_2⟩
               -> model(⟨emb_vec_1, emb_vec_2⟩)
               -> ⟨emb_a, emb_b, emb_c SEP emb_d, emb_e SEP emb_i, emb_j, emb_k, emb_l⟩
```

Given the two-file diff, the model returns a hierarchical structure of which tests to run. That also lets us parametrize the CI workload using a budget: if the budget allows, we also execute the tests defined in tier B, and if there's still budget (or the reviewer chooses to), we run tier C too. This also gives the reviewer a picture of all the tests that should be running vs. the minimum set we're going to launch as soon as the ready label is added. It makes the selective heuristic more observable and controllable.

So to sum up:

- Observability is the blocker for the pure LLM approach because it'll work until it doesn't, and we won't know which.
- Two mitigations on the agentic side: force it to log the candidate test set and how reasoning narrowed it, and refresh its memory frequently so we incrementally build the dependency graph.
- Staleness is a shared problem for any heuristic. We need a story for how the model/memory stays current after large PRs, not just at setup time.
- The embedding-based alternative trades some capability for a lot of control: LLM only generates per-function/module embeddings, a trained scoring model maps diff -> tiered test sets (A/B/C).
- Budget-parametrized execution: tier A always runs on ready, B and C are opt-in based on CI budget or reviewer choice.
- Reviewer gets visibility into the full set of tests the model thinks are relevant vs. the minimum set actually launched and hence the heuristic becomes controllable instead of a black box.

I have not tried yet to train a model with the embeddings idea. But we can go a bit in parallel here with the different efforts and ideas.

**Comment by tarukumar** [comment]
These are the Test Selection Strategies I had listed earlier.

## Test Selection Strategies

---

###  Solution A: Dependency-Aware Selection (Deterministic)

Uses **dynamic analysis** to map which tests touch which code.

**Plan**
- Tooling: `pytest-testmon`, `coverage.py`
- Mechanism: Records execution traces → runs only affected tests

**Steps**
```bash
pip install pytest-testmon
pytest --testmon   # baseline
pytest --testmon   # incremental
```
Pros

✅ High precision (runs only required tests)
✅ No training/setup time needed
✅ Deterministic and reliable

Cons

❌ Requires caching .testmondata
❌ Slight runtime overhead due to tracing
❌ State can become stale if not maintained

Solution B: Predictive Test Selection (ML-based)

Uses ML to estimate probability of failure (P_f).

**Plan**
Model considers:

- Code churn
- Historical failures
- Author patterns
- Commit timing

**Steps**

1. Record test data (2–4 weeks)
2. Train model
3. Select subset in CI
4. e.g. 20% tests → 90% confidence

Pros

✅ Maximum time savings (~80%+)
✅ Learns hidden dependencies
✅ Improves over time

Cons

❌ Risk of missing failing tests
❌ Requires training data + infra
❌ Harder to debug (black-box)

Solution C: Heuristic-Based Selection (Rule-Based)

Manual mapping between code and tests.

**Plan**

- Tooling: Python scripts + pytest markers
- Mechanism: File path → test mapping

**Step**

1. git diff
2. pytest <relevant_tests>

Pros

✅ Simple and transparent
✅ No external tools required
✅ Easy to implement quickly

Cons

❌ Manual maintenance required
❌ Breaks when structure changes
❌ Misses indirect dependencies

**Comment by tarukumar** [comment]
Solution C (heuristics, paths, markers): This is similar to what is proposed earlier (without agentic)  git diff → map → pytest paths.This could be the baseline.  The issue is drift and incomplete maps.

Add A selectively: e.g. one premerge job or dev flow where `pytest --testmon (or coverage-diff)` is realistic, with a well-defined cache key and a fallback to full when DB missing/stale.

B later(Pure  LLM): for scheduling, budgeting, or risk - with clear “never skip required tiers” policy.

Based on what I can see, there are still missing issues that none of the current solutions address alone:

- Flaky tests and infrastructure failures  when environments differ.
- C++/CUDA and other non-Python changes are not fully covered - A/C require explicit rules, and currently this is handled only through run_all_patterns or full runs.

**Comment by tarukumar** [comment]
1. https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/45880.pdf 
2. https://software-lab.org/publications/issta2021_TestFailure.pdf 
3. https://www.testmon.org/

**Comment by edwardquarm** [comment]
Thanks for putting the design docs together. Since we want to get something practical up and running, I do think the agentic approach is a reasonable place to start.

I also like the idea of keeping the initial trigger the same as you described in the doc: PR opened or updated, collect the diff, generate the source→test mapping, load the static rules, and then let the agent reason over that context before posting the result back as a PR comment in shadow mode.

The part I think is worth refining is the first agent output. Instead of asking the agent to return a flat final list of tests, maybe we should ask it to first build a candidate set from the mapping and rules, and then organize that set in a way that helps us fail fast.

Maybe we should think about the output as an ordered tree with two top-level groups: **direct** and **indirect**.

```text
Candidate tests
├── Direct
│   ├── Tests that directly exercise the changed code path
│   ├── Tests that import or invoke the changed module directly
│   └── Tests that cover the exact entrypoint affected by the PR
│
└── Indirect
    ├── Tests that cover the same subsystem
    ├── Tests that exercise the same runtime path through fixtures/helpers/subprocesses
    └── Critical regression / safety tests for that area
```

The idea would be to run the **Direct** group first, since those are the tests most likely to fail if the change introduces a regression, and then run the **Indirect** group next for broader confidence.

That feels easier to reason about than having the agent return a flat list, and it also makes the behavior more observable.

**Comment by khluu** [comment]
Thanks all for your thoughtful comments/ideas!

To start simple, I think we can go with shadow mode in @avinashsingh77 's design first. Some of my thoughts on the comments:
- Observability: we can have a bot that comments on PRs to list tests that should run and reasoning behind it. This would help collect feedbacks from the community and also track the performance of the bot.
- Staleness/memory: we can have the LLM take in the codebase as context for now, then we can optimize by caching/refreshing this periodically. The cost to have LLM consume this context would be significantly less than the amount saved by not running extra tests.
-  Running direct/indirect groups: I think this makes sense but I'd put this as P1

We can discuss more about logistics/collaboration in https://vllm-dev.slack.com/archives/C07R5PAL2L9/p1776191122982049
